### DISCUSSION TITLE
Chord Identification using DryWetMIDI

### DISCUSSION THREAD

**Discussion by Jenix-Park** [melanchall/drywetmidi#260] [Q&A] [answered]
Hello,

I am planning to incorporate a functionality that allows users to identify a chord name with two or more given notes.
To test this feature, I ran the following code:

```
var chord1 = new Melanchall.DryWetMidi.MusicTheory.Chord(NoteName.C, NoteName.E, NoteName.G);
var chord2 = new Melanchall.DryWetMidi.MusicTheory.Chord(NoteName.E, NoteName.C, NoteName.G);

foreach (var chordName in chord1.GetNames())
    Console.WriteLine(chordName);

Console.WriteLine();

foreach (var chordName in chord2.GetNames())
    Console.WriteLine(chordName);
```

The outputs I received were as follows:

```
C
CM
Cmaj
C5/E

C
CM
C5/E
Cmaj
```

However, I would prefer the following:

1. Instead of suggesting multiple possible chords, I would like to print the most suitable chord name.
In the mentioned case, the desired results would be "C/E" and "Cmaj" respectively.
(At least, I can live with the "C5/E" output in the first case, though)

2. It would be convenient if I could directly pass note numbers rather than note names.
My application utilizes note numbers for various tasks, and converting or creating new Note objects could result in unnecessary garbage collection and overhead.
Performance is a primary concern for me, so I would like to avoid any additional layers if possible.

Is it possible to achieve these with the library?
If not, do you know any other library for this specific case?
ChatGPT recommended NAudio, but it doesn't seem to be a proper one.

I greatly appreciate your plugin, which has been of tremendous help to me.

Thank you for your ongoing support.

**Comment by melanchall** [comment, marked as answer]
Hello,

1. First of all, it's required to return multiple names due to the simple thing: different people name the same chord differently :-) As for most suitable names I suppose you mean `C/E` for second chord (`E C G`) and `Cmaj` (`C E G`) for the first one. Even here we can see that you prefer to name the first chord `Cmaj` but there are many people who will name it just `C`.

    But your idea is interesting. Hmm... Placing a name where root note matches a chord's root note to the first place in a result. I should think but now that's not possible. Names are sorted by their lengths.
2. A music theory chord is a set of notes without octaves so I think the current way is a proper one. But you can easily get names from numbers:
    ```csharp
    var noteNames = noteNumbers
        .Select(n => NoteUtilities.GetNoteName((SevenBitNumber)n))
        .ToArray();
    ```

> If not, do you know any other library for this specific case?
> ChatGPT recommended NAudio, but it doesn't seem to be a proper one.

I don't know such libraries for  .NET. You can try to search for libraries for other languages. NAudio is a great library, but it's not for MIDI, it's for audio. Yes, it has MIDI API, but it's super low level and exists only to exist (the author of the library said that directly in one of the issues).

**Reply by Jenix-Park** [reply]
DryWetMIDI is hands down the absolute best when it comes to MIDI features!
I can't imagine my music journey without it, and I have no desire to switch to any other plugins.

By the way, I was actually excited about the chord identification feature in DryWetMIDI since I thought it was already available.
I totally get what you said, but I was just hoping there was a way to provide users with a friendly guidance, just like those Casio keyboards and Logic Pro do, by showing a single chord.

But hey, no worries!
I understand that the final decision is up to me, and I'll take the time to consider my options and pick the one that suits me best.

Oh, and about NAudio?
Honestly, I don't know much about it.
I was just curious to see if it had any chord identification capabilities.

I want to sincerely thank you for your super quick and kind response.
Your help has been beyond amazing and incredibly appreciated! You rock!

**Reply by Jenix-Park** [reply]
@melanchall Oh, by the way, what will happen if all twelve or random notes are passed to the Chord class and the GetNames() method is called?
My main concern here is whether it is acceptable to directly pass all the notes inputted by a user to the Chord class, without requiring my app to specifically identify and ignore those cases.

**Reply by melanchall** [reply]
Thank you for such kind words!

> I totally get what you said, but I was just hoping there was a way to provide users with a friendly guidance, just like those Casio keyboards and Logic Pro do, by showing a single chord.

I'll think about adding some parameter to the `GetNames` method to specify sorting algorithm, but unfortunately that won't be in an upcoming release. Anyway thank you for your suggestion.

> Oh, by the way, what will happen if all twelve or random notes are passed to the Chord class and the GetNames() method is called?

`GetNames` will return an empty collection because DryWetMIDI doesn't know such a chord. The library uses [Chords names table](https://melanchall.github.io/drywetmidi/articles/music-theory/Chord.html#chords-names-table) (CNT) to identify a chord's name so you can follow the link and see what chords are currently can be recognized. DryWetMIDI uses CNT instead of some algorithmic solution to provide best performance of the operation.

**Reply by Jenix-Park** [reply]
Then, I guess I can simply throw all the notes into the Chord class and see if it returns an empty collection, haha. I'm not sure about the performance efficiency, but at least the code will be much simpler.

Thanks a lot!

**Comment by Jenix-Park** [comment]
@melanchall Sorry, I have one final question.
This time, I have decided to consistently exclude slash chords.
In the previous example, I will choose between C, CM, and Cmaj, but not C5/E.

Is there a possibility that only one slash chord will be returned?
I ask this because I plan to disregard any chords that include the '/' character. This could be problematic if there is a scenario where only one slash chord is returned.

**Reply by melanchall** [reply]
> Is there a possibility that only one slash chord will be returned?

It's hard to say but I think that's possible in theory.

To clarify, if these names returned

```
X
X/Y
Z
```

you need to leave only

```
X
Z
```

right?

But if only a slash chord returned

```
X/Y
```

return it, yes?

If so, just filter out strings containing `/`. If you get empty collection after filtering, original one contains only slash chords, so just pick up any one you want from the original collection.

**Reply by Jenix-Park** [reply]
Got it! Thank you!

**Reply by Jenix-Park** [reply]
@melanchall 
Hello again!

I hope you're doing well.
I've been experimenting with the chord name identifier feature in a new project using the method we discussed earlier.
However, I ran into a bit of a hiccup with the code snippet below:


```
var chord = new Melanchall.DryWetMidi.MusicTheory.Chord(noteNameList);
var chordNames = chord.GetNames();  // This line is the problem!
```


As we chatted about earlier, I initially planned to feed all the notes from user input into the Chord constructor, thinking that GetNames() would handle things seamlessly.
To my surprise, I noticed that when working with 8 notes, GetNames() started to slow down, and with 10 notes, it seemed like IT WAS TAKING AGES!

Now, it's crucial to emphasize that my app is a 'real-time performance instrument' app, and it needs to run smoothly even on low-end mobile devices.
The code above was tested in an empty project on a high-end PC with the 12th Gen Intel i5-12600K, which is more than fast enough.
My main concern is ensuring that this feature doesn't hinder performance on less powerful devices.

So, given the real-time nature of my application and the variation in device performance, I'd love to hear your thoughts on these questions:

1. Could you share your insights on whether calling GetNames() on a Chord object with just 4 notes might introduce some degree of latency, particularly on low-end mobile devices?

2. If there's no added latency with 4 notes on high-end hardware, could you give me a rough idea of how many notes I can safely add to a Chord object without stirring up latency issues, especially on less powerful devices?

I'm genuinely thankful for your insights.
Thanks a bunch!

**Reply by melanchall** [reply]
Hi @Jenix-Park,

Yes, that's because `ChordsNamesTable` calculates permutations of the notes collection to collect all possible names of a chord. The number of all possible permutations for `N` elements is `N!` (factorial) which quickly gives big numbers for `N` (for `N = 8` we'll get `40320` possible variants). And then for each permutation we need to go through the table of notes offsets to determine a name.

I have no idea how we can avoid permutation complexity, but maybe we can speed up each permutation check time. As an option we can have additional parameter for the `GetNames` method which tells the engine some optimization:

```csharp
var names = chord.GetNames(ChordNameSearchHint.SkipInversions);
```

But inversions are important so I think this is not an option. I need to think how optimize this part of the algorithm.

Anyway, what are such chords that contain more than 5 notes? Are those real ones or you just have no any filtering? I suggest to limit a chord's length by, for example, 5 notes. Then you can receive users feedback whether this is OK or not.

> Could you share your insights on whether calling GetNames() on a Chord object with just 4 notes might introduce some degree of latency, particularly on low-end mobile devices?

4 notes shouldn't create the problem. If you can limit a chord's length by 4 notes, it would be even better.

**Reply by Jenix-Park** [reply]
@melanchall 
Thank you so much for your help!

In my previous test, I used note numbers 1 through 10 and then converted them using NoteUtilities.GetNoteName(), just as you had suggested. I was initially thinking of passing all the notes from user input without any filtering. However, after some testing, I realized that I needed to limit the number of notes.

At first glance, 7 notes seemed to work well on my PC, but when I timed all the involved code with the Stopwatch class, I noticed that even just one note introduced some level of latency.

As far as I understand, latency for real-time instruments used by professional musicians should be 15ms or below. My app currently has around 15ms of latency, so adding another 1ms would be a significant issue.

Because of this, I've decided to put the chord identifier feature on hold for now and focus on reducing latency. Once I've optimized the performance, I'll revisit the idea of working with 5 or fewer notes.

Once again, I want to express my gratitude for your assistance!

**Reply by melanchall** [reply]
@Jenix-Park Just to let you know: I have some ideas to simplify the algorithm and increase its performance. I'll write you when I get tests/benchmarks results.

**Reply by Jenix-Park** [reply]
Wow, great to hear!
Thanks for your hard work!

**Reply by melanchall** [reply]
@Jenix-Park I've built the [Melanchall.DryWetMidi-7.0.2-prerelease7](https://www.nuget.org/packages/Melanchall.DryWetMidi/7.0.2-prerelease7) version. Can you please try it and let me know how fast it is? My benchmarks show significant speed up.

Also the algorithm will exit immediately now if notes count is greater than supported by the algorithm (in this case we know for sure that we won't find any name so there is no need to run detection routine at all). More than that, implementation is now simpler and doesn't use permutations.

Also I found a bug in `9sus2` chords detection :-)

**Reply by Jenix-Park** [reply]
@melanchall 
WOW!

I’m absolutely thrilled with your new code version—it’s like lightning in a bottle! It’s so quick that I can barely believe it doesn’t add even a smidge of latency, no matter how many notes I throw at it. 😄
(Just a quick heads up, I whipped up my test code pretty quickly, so there might be a few hiccups there. Plus, I ran it on my PC, so things might be a tad different on less powerful devices.)

I’ve got a couple of questions on my mind:

1. Does this awesome update mean I can send all the user input straight through without having to filter it first?
2. Can it handle heaps of notes, like 9 or 10?
3. I’m curious if there’s any way to tweak my **GetChordNames()** method below to make it even more efficient—any thoughts on minimizing overhead from things like garbage collection, memory allocation, or method calls?

Thanks a million for your amazing work and your time!

```
class TestProgram
{
    const int TEST_COUNT = 500000;

    const int MIN_USERINPUT = 3;    // 0 ~ 88 (Minimum count of piano keys pressed by the user)
    const int MAX_USERINPUT = 12;


    static List<NoteName> noteNameListToCheck = new List<NoteName>(88);

    static IReadOnlyCollection<string> GetChordNames(List<int> noteNumbers)
    {
        // This removes duplicates of note names across different octaves, treating them as a single note.
        var uniqueNoteNumbers = noteNumbers
            .Select(n => n % 12)
            .Distinct();

        if (uniqueNoteNumbers.Count() < 2)
        {
            noteNameListToCheck.Clear();

            return Array.Empty<string>();
        }

        foreach (var noteNumber in uniqueNoteNumbers)
            noteNameListToCheck.Add(NoteUtilities.GetNoteName((SevenBitNumber)noteNumber));

        var chordNames = new Chord(noteNameListToCheck).GetNames();

        noteNameListToCheck.Clear();

        return chordNames;
    }


    static Random randomObj = new Random();
    static HashSet<int> uniqueNumbers = new HashSet<int>();

    static void SimulateUserInput(List<int> list)
    {
        var elementCount = randomObj.Next(MIN_USERINPUT, MAX_USERINPUT + 1);

        while (uniqueNumbers.Count < elementCount)
        {
            var number = randomObj.Next(1, 89);
            uniqueNumbers.Add(number);
        }

        list.AddRange(uniqueNumbers);

        uniqueNumbers.Clear();
    }


    static void Main(string[] args)
    {
        List<int> userInput = new List<int>(88);

        Stopwatch sw = new Stopwatch();

        for (int i = 0; i < TEST_COUNT; ++i)
        {
            userInput.Clear();

            SimulateUserInput(userInput);

            // Temporarily disabling this to test with simulated user input.
            //if (userInput.Count < 2)
            //    continue;

            sw.Start();

            var chordNames = GetChordNames(userInput);

            sw.Stop();

            if (chordNames.Count > 0)
            {
                foreach (var chordName in chordNames)
                    Console.Write(chordName + ", ");

                Console.WriteLine();
            }
        }

        Console.WriteLine("----------------------------------------------------------");
        Console.WriteLine("Total : " + sw.Elapsed.TotalMilliseconds);
        Console.WriteLine("Average : " + sw.Elapsed.TotalMilliseconds / TEST_COUNT);
    }
}
```

**Reply by melanchall** [reply]
@Jenix-Park Great news!

> Does this awesome update mean I can send all the user input straight through without having to filter it first?

Yes. All required filtering will be applied within the `GetNames` method.

> Can it handle heaps of notes, like 9 or 10?

Yes. With a lot of notes the method will just exit immediately because there is no rules (in the library) for such chords how to name them.

> I’m curious if there’s any way to tweak my GetChordNames() method below to make it even more efficient—any thoughts on minimizing overhead from things like garbage collection, memory allocation, or method calls?

Hmm, your method is already simple, so I don't see obvious ways to improve the code.

Thank you for your questions and issues!

**Reply by Jenix-Park** [reply]
@melanchall I just gave the old version a spin with my test code from above, and wow, what a difference!
It was pretty much a no-go.
The new version, on the other hand, is like night and day—it's amazing!
Can't wait for the official release to drop.

Thanks a million for all your hard work and for getting back to me. You rock! :)

**Reply by melanchall** [reply]
@Jenix-Park Thank you! Unfortunately I can't provide you any date when the release will happen :-(

**Comment by Tony-Pappas** [comment]
Hey guys, you may want to look at Adrian Akison's 2008 article (and the demo code) for permutations. The performance is ripping fast. https://www.codeproject.com/Articles/26050/Permutations-Combinations-and-Variations-using-C-G

**Reply by melanchall** [reply]
In fact, there is no need in permutations in the chord name detection algorithm. But thank you anyway :-)

**Comment by valimaties** [comment]
I love to see this feature.
I don't know if any of you play arrangers (Yamaha, Korg, etc). However, I have a Yamaha Genos and also a Korg PA5X. These ones have some "advanced" modes for detecting chords based on note you played in the arrangement section. The best one from these two ones is Genos, which AI mode detection is very very well done. 
For example, when you play only G-E this will be translated into a Em/G chord, so a Emin with a G in bass (last picture).
I will attach some pictures to see how the keyboard recognized those "2-keys" chords:

C/E
![C-E](https://github.com/melanchall/drywetmidi/assets/49796453/28202b77-f095-4286-a329-24f17691b8d2)

C/G
![C-G](https://github.com/melanchall/drywetmidi/assets/49796453/84c8b27d-4b5c-4517-b41b-c46b61ffd620)

Em/G
![Em-G](https://github.com/melanchall/drywetmidi/assets/49796453/d59b9aaa-5528-4c8a-b0ea-e1c2c72175ca)

**Reply by melanchall** [reply]
Hi,

Interesting feature to guess a chord by two notes. But I have questions. Let's see at `G E`. Your device recognized this sequence as `Em/G`. But why not `C`, for example? Both `Em` and `C` chords have `G` and `E` notes, so if we are agreed to ignore a third note, we have in fact several variants, havent't we?

**Reply by valimaties** [reply]
`Em` is composed of `E-G-B` notes. When I have only two notes involved, like `E` and `G` , the heigher note gives me the chord (`E`) and the second note gives me the bass note and in the same time tells keyboard what type of chord is, `minor` or `major`. Because `G` is the `minor 3rd` of the `E` chord then the result is a `Em/G` . If the `E` had been in the same octave (the heigher key had been `G`) the keyboard had been translate it as a `Em` chord, simple. 
For the case of `GC`, the higher note is `C`, which compose the `C` chord. The `G` is lower note, so tells keyboard that it must be the bass note as well, and in the same time is the fifth note in the chord, no `minor 3rd` or `major 3rd` involved, so the keyboard decide that is a major chord. 

For the `Em/G` case, if instead of `G` I take `G#` it will be a `E/G#` :) Nice, isn't it?

PS: Chords like `X/Y` will be composed like this only if X key is in the next octave!

**Reply by melanchall** [reply]
@valimaties Thanks for the explanation! I'll think about how to integrate similar logic into the library.

**Reply by valimaties** [reply]
Be aware of those chords created with two notes, when notes belong to the same octave. These are in strong bound with the `Chord Fingering` type in the arranger. The above chord recognition is available only in `AI fingered` mode on Yamaha and `Expert`(`Advanced`) mode on Korg. If the `Chord Fingering` is other than those two, the recognition will be changed accordingly.
If for example `Chord Fingering` is `Single finger` - `A#(2) C3` (2 and 3 are octave no) is not recognized as a chord with specific bass note, but is recognized as `Cm`, `B C` is recognized as `C7`, as well as the three notes like ` A# B C` (in this order) is recognized as a `Cm7`. 

Now, if in `Single finger` the `Cm` chord is recognized using `A# C` notes in this order, in `AI mode` the `Cm` chord is the `C D#` notes, in the same octave.

So, in arrangers world those two-notes chords having different bass note are recognized only in those `Chord Fingering` "special (advanced)" modes.

Regards,
Vali
